### PR TITLE
Add FreeVC implementation

### PR DESCRIPTION
**Pull request by Nugine**
## ✨ Description

**FreeVC: Towards High-Quality Text-Free One-Shot Voice Conversion**
+ https://arxiv.org/abs/2210.15418
+ https://github.com/OlaWod/FreeVC

This PR is a part of AIR6063 final project.
+ @Nugine (223040051)
+ @SeanYouLaw (223040034)

## 🚧 Related Issues

During the project, we also open some issues and another PR to help improve Amphion.
+ #195 
+ #196 
+ #197 
+ #198 

## 👨‍💻 Changes Proposed

- [x] Add FreeVC in models
- [x] Add FreeVC in egs

## 🧑‍🤝‍🧑 Who Can Review?

[Please use the '@' symbol to mention any community member who is free to review the PR once the tests have passed. Feel free to tag members or contributors who might be interested in your PR.]
@zhizhengwu @RMSnow @Adorable-Qin

## ✅ Checklist

- [ ]  Code has been reviewed
- [ ]  Code complies with the project's code standards and best practices
- [ ]  Code has passed all tests
- [ ]  Code does not affect the normal use of existing features
- [ ]  Code has been commented properly
- [ ]  Documentation has been updated (if applicable)
- [ ]  Demo/checkpoint has been attached (if applicable)

### CODE REVIEW

**Comment by SeanYouLaw** [comment]
Here are some examples of our results:

https://github.com/open-mmlab/Amphion/assets/58773169/3256156a-d77a-4079-a437-6b163a8d0c12
https://github.com/open-mmlab/Amphion/assets/58773169/02a1e70e-ec1b-4755-b4e5-9e50ebdb4620
https://github.com/open-mmlab/Amphion/assets/58773169/cac3e57b-782e-43f7-96d5-48420c113f23

---
https://github.com/open-mmlab/Amphion/assets/58773169/0bec78cd-1543-44dd-9673-9842dd96f7fa
https://github.com/open-mmlab/Amphion/assets/58773169/a7faf334-d863-4d11-beb5-4bd220055a7b
https://github.com/open-mmlab/Amphion/assets/58773169/055a3d9b-572f-4079-811a-f15becbc2c83

---
https://github.com/open-mmlab/Amphion/assets/58773169/2fc73678-0e60-4665-89ce-52c96aeeaded
https://github.com/open-mmlab/Amphion/assets/58773169/56dec48f-cc1d-403c-af30-ef80b29343a7
https://github.com/open-mmlab/Amphion/assets/58773169/fa8e6091-70f0-431f-b809-82c4c9bee39d

**Comment by RMSnow** [comment]
The quality of the samples sounds good. @Adorable-Qin Please check the code and document carefully.

**Comment by SeanYouLaw** [comment]
Here are some examples of our results, using the checkpoint of around 200 epoch training(while above examples are from the pretrained checkpoint):

https://github.com/open-mmlab/Amphion/assets/58773169/3b85076d-187c-4329-9405-f71eea734b89


https://github.com/open-mmlab/Amphion/assets/58773169/b3a1f57a-7e56-4ccc-a319-f7f22bdf11f5


https://github.com/open-mmlab/Amphion/assets/58773169/bbde9081-e414-41e8-848b-304176d871cd


https://github.com/open-mmlab/Amphion/assets/58773169/6754c030-5703-4b5f-afce-85e49282eb55

https://github.com/open-mmlab/Amphion/assets/58773169/0f7b0bc7-8b03-456e-9f59-4ba1b13be955


https://github.com/open-mmlab/Amphion/assets/58773169/de76a1da-95df-4aab-a727-1ccd24bcaaf0


https://github.com/open-mmlab/Amphion/assets/58773169/d849b433-5478-46ca-ae85-f9737bf6de57

https://github.com/open-mmlab/Amphion/assets/58773169/0e1dfd5b-c09f-4da9-842b-94acb49c2bdc


https://github.com/open-mmlab/Amphion/assets/58773169/6ae3b513-035c-42fb-8f0f-cf5041098850

**Comment by Nugine** [comment]
Our AutoDL server will expire tomorrow. Here is a demo video recording the training status.

https://github.com/open-mmlab/Amphion/assets/30099658/06e43935-bf2d-4930-90e9-c89642b623cd

File: models/vc/FreeVC/data.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+        lengths = []
+        for audiopath in self.audiopaths:
+            path = os.path.join(self.vctk_16k_dir, audiopath)
+            lengths.append(os.path.getsize(path) // (2 * self.hop_length))
+        self.lengths = lengths
+
+    @torch.no_grad()
+    def load_sample(self, filename):
+        filepath = os.path.join(self.vctk_16k_dir, filename)
```
Comment (Adorable-Qin) [inline models/vc/FreeVC/data.py:58]: Why is this line hard-coded? Is it possible to select datasets in config?

File: models/vc/FreeVC/hifigan.py
Comment (Adorable-Qin) [inline models/vc/FreeVC/hifigan.py:1]: It would be great if you could check if this part could reuse the HiFiGAN provided in the vocoder part of Amphion later.

File: models/vc/FreeVC/mel_processing.py
Comment (Adorable-Qin) [inline models/vc/FreeVC/mel_processing.py:1]: Some functions are duplicated, check: `utils/mel.py`

File: models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt
Comment (Adorable-Qin) [inline models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt:1]: What's the purpose of this file?

File: models/vc/FreeVC/speaker_encoder/data_objects/speaker_verification_dataset.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if len(speaker_dirs) == 0:
+            raise Exception(
+                "No speakers found. Make sure you are pointing to the directory "
+                "containing all preprocessed speaker directories."
+            )
+        self.speakers = [Speaker(speaker_dir) for speaker_dir in speaker_dirs]
+        self.speaker_cycler = RandomCycler(self.speakers)
+
+    def __len__(self):
+        return int(1e10)
```
Comment (Adorable-Qin) [inline models/vc/FreeVC/speaker_encoder/data_objects/speaker_verification_dataset.py:24]: Why is this length hard-coded?

File: models/vc/FreeVC/speaker_encoder/hparams.py
Comment (Adorable-Qin) [inline models/vc/FreeVC/speaker_encoder/hparams.py:1]: It would be better to change the hparams into an Amphion-style config file.

File: models/vc/FreeVC/speaker_encoder/train.py
Comment (Adorable-Qin) [inline models/vc/FreeVC/speaker_encoder/train.py:1]: Is it possible to support multi-GPU training using external library like the [Accelerate](https://huggingface.co/docs/accelerate/index) used in Amphion?

File: models/vc/FreeVC/train.py
Comment (Adorable-Qin) [inline models/vc/FreeVC/train.py:1]: Is it possible to support multi-GPU training using external library like the [Accelerate](https://huggingface.co/docs/accelerate/index) used in Amphion?

File: models/vc/FreeVC/data.py (hunk 1 above)
Comment (Nugine) [reply models/vc/FreeVC/data.py:58]: The original implementation only trains on VCTK dataset. 

+ Data preprocessing uses the file structure of VCTK dataset to retrieve speaker tags.
+ When splitting train/val/test set, every speaker's samples are split randomly. It ensures that every speaker is in train & val & test set.

It's possible to support other datasets if we can perform the same operations on them.

Comment (Adorable-Qin) [reply models/vc/FreeVC/data.py:58]: Thank you for your explanation! 
@RMSnow For this implementation, do we expect a universal model that can be trained on any dataset?

File: models/vc/FreeVC/hifigan.py
Comment (Nugine) [reply models/vc/FreeVC/hifigan.py:1]: Changed in [74203eb](https://github.com/open-mmlab/Amphion/pull/201/commits/74203ebe58c9844537eed71403e92f570ffb42ea)

File: models/vc/FreeVC/mel_processing.py
Comment (Nugine) [reply models/vc/FreeVC/mel_processing.py:1]: Some functions are not exactly the same with those in `utils/mel.py`.
So I'm not sure which version I should adopt, the original implementation or Amphion implementation?

File: models/vc/FreeVC/hifigan.py
Comment (Adorable-Qin) [reply models/vc/FreeVC/hifigan.py:1]: Appreciate your help!

File: models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt
Comment (Nugine) [reply models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt:1]: The whole directory (`models/vc/FreeVC/speaker_encoder`) is copied from
+ https://github.com/OlaWod/FreeVC/tree/81c169cdbfc97ff07ee2f501e9b88d543fc46126/speaker_encoder  
(MIT license)
+ https://github.com/liusongxiang/ppg-vc/tree/b59cb9862cf4b82a3bdb589950e25cab85fc9b03/speaker_encoder  
(Apache-2.0 license)

We keep it unchanged to match the original implementation.
However, it may be a problem if we copy so much code and a pretrained ckpt from other repo. I'm not sure what is the best practice.

File: models/vc/FreeVC/speaker_encoder/data_objects/speaker_verification_dataset.py (hunk 2 above)
Comment (Nugine) [reply models/vc/FreeVC/speaker_encoder/data_objects/speaker_verification_dataset.py:24]: Same with above. The code is from other repo.

File: models/vc/FreeVC/speaker_encoder/hparams.py
Comment (Nugine) [reply models/vc/FreeVC/speaker_encoder/hparams.py:1]: Same with above. The code is from other repo.

File: models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt
Comment (Adorable-Qin) [reply models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt:1]: Thank you for your explanation.
@RMSnow Any advice about this?

File: models/vc/FreeVC/mel_processing.py
Comment (Nugine) [reply models/vc/FreeVC/mel_processing.py:1]: Changed in [d42e665](https://github.com/open-mmlab/Amphion/pull/201/commits/d42e665c9d61fd8efe28b567022608975e82f2bc).

File: models/vc/FreeVC/data.py (hunk 1 above)
Comment (RMSnow) [reply models/vc/FreeVC/data.py:58]: Yes, a universal FreeVC for any datasets are welcome. I think only the FreeVC's model part need to be integrated in Amphion.

File: models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt
Comment (RMSnow) [reply models/vc/FreeVC/speaker_encoder/ckpt/pretrained_bak_5805000.pt.txt:1]: @Adorable-Qin I think introducing such a pretrained speaker encode is acceptable. It is just like WeNet. However, please add some acknowledge in our main readme before integrating it.

BTW, I think the `.pt.txt` file is strange. If it is a pretrained model, we can follow our [pretrained model's part](https://github.com/open-mmlab/Amphion/tree/main/pretrained) to integrate.
